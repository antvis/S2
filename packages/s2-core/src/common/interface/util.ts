export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
