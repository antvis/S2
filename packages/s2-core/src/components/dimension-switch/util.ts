import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
