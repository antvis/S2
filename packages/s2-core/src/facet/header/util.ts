import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
