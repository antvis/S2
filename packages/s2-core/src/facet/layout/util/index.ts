export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
