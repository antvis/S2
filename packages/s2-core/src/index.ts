export { Event as GEvent } from '@antv/g-canvas';
export { getTheme } from './theme';
export { Node } from './facet/layout/node';
export { Hierarchy } from './facet/layout/hierarchy';
export { buildTableHierarchy } from './facet/layout/build-table-hierarchy';
export * from './common';
export * from './utils';
export * from './cell';
export * from './sheet-type';
export * from './data-set';
export * from './interaction';
export * from './ui/tooltip';
export * from './facet/header';
