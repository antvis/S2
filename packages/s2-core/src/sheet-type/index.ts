import { PivotSheet } from './pivot-sheet';
import { TableSheet } from './table-sheet';
import { SpreadSheet } from './spread-sheet';

export { PivotSheet, TableSheet, SpreadSheet };
