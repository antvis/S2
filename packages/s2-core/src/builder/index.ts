import { deepMix } from '@antv/util';
import {
  CellCallback,
  Conditions,
  CornerHeaderCallback,
  DataCellCallback,
  FrameCallback,
  HierarchyCallback,
  KeepOnlyIds,
  LayoutArrangeCallback,
  LayoutCallback,
  LayoutResultCallback,
  NodeField,
  Pagination,
  SpreadsheetOptions,
  CustomHeaderCells,
  Style,
  Totals,
} from '../common/interface';
import { TREE_ROW_DEFAULT_WIDTH } from '../common/constant';
import { DefaultStyleCfg } from './default-style-cfg';

export class SpreadSheetOptionsBuilder {
  public static Builder(): SpreadSheetOptionsBuilder {
    return new SpreadSheetOptionsBuilder();
  }

  private _conditions: Conditions;

  private _width: number;

  private _height: number;

  private _hierarchyCollapse: boolean;

  private _hierarchyType: 'grid' | 'tree';

  private _linkFieldIds: string[];

  private _pagination: Pagination;

  private _containsRowHeader: boolean;

  private _spreadsheetType: boolean;

  private _style: Partial<Style>;

  private _totals: Totals;

  private _customHeaderCells: CustomHeaderCells;

  private _showSeriesNumber: boolean;

  private _hideNodesIds: string[];

  private _keepOnlyNodesIds: KeepOnlyIds;

  private _registerDefaultInteractions: boolean;

  private _scrollReachNodeField: NodeField;

  private _hideRowColFields: string[];

  private _valueInCols = true;

  // 自定义单元格cell
  private _dataCell: DataCellCallback;

  // 自定义cornerCell
  private _cornerCell: CellCallback;

  // 自定义行头cell
  private _rowCell: CellCallback;

  // 自定义列头cell
  private _colCell: CellCallback;

  // 自定义 frame 边框
  private _frame: FrameCallback;

  // 角头可能需要全部自定义，而不是用交叉表概念的node来渲染
  private _cornerHeader: CornerHeaderCallback;

  // 自定义layout
  private _layout: LayoutCallback;

  // 自定义布局结果
  private _layoutResult: LayoutResultCallback;

  // 行列结构的自定义
  private _hierarchy: HierarchyCallback;

  // 行列 维度值的自定义排序
  private _layoutArrange: LayoutArrangeCallback;

  private _debug: boolean;

  public build(): SpreadsheetOptions {
    return {
      debug: this._debug,
      conditions: this._conditions,
      width: this._width,
      height: this._height,
      hierarchyCollapse: this._hierarchyCollapse || false,
      hierarchyType: this._hierarchyType || 'grid',
      linkFieldIds: this._linkFieldIds || [],
      pagination: this._pagination,
      containsRowHeader: this._containsRowHeader || false,
      spreadsheetType: this._spreadsheetType,
      style: deepMix({}, DefaultStyleCfg(), this._style),
      totals: this._totals,
      customHeaderCells: this._customHeaderCells,
      showSeriesNumber: this._showSeriesNumber || false,
      hideNodesIds: this._hideNodesIds || [],
      keepOnlyNodesIds: this._keepOnlyNodesIds || { rowIds: [], colIds: [] },
      registerDefaultInteractions: this._registerDefaultInteractions,
      scrollReachNodeField: this._scrollReachNodeField || {
        rowField: '',
        colField: '',
      },
      hideRowColFields: this._hideRowColFields || [],
      valueInCols: this._valueInCols || false,
      dataCell: this._dataCell,
      cornerCell: this._cornerCell,
      rowCell: this._rowCell,
      colCell: this._colCell,
      frame: this._frame,
      cornerHeader: this._cornerHeader,
      layout: this._layout,
      layoutResult: this._layoutResult,
      hierarchy: this._hierarchy,
      layoutArrange: this._layoutArrange,
    } as SpreadsheetOptions;
  }

  public debug(debug: boolean) {
    this._debug = debug;
    return this;
  }

  public layoutArrange(layoutArrange: LayoutArrangeCallback) {
    this._layoutArrange = layoutArrange;
    return this;
  }

  public hierarchy(hierarchy: HierarchyCallback) {
    this._hierarchy = hierarchy;
    return this;
  }

  public layoutResult(layoutResult: LayoutResultCallback) {
    this._layoutResult = layoutResult;
    return this;
  }

  public layout(layout: LayoutCallback) {
    this._layout = layout;
    return this;
  }

  public cornerHeader(cornerHeader: CornerHeaderCallback) {
    this._cornerHeader = cornerHeader;
    return this;
  }

  public frame(frame: FrameCallback) {
    this._frame = frame;
    return this;
  }

  public colCell(colCell: CellCallback) {
    this._colCell = colCell;
    return this;
  }

  public rowCell(rowCell: CellCallback) {
    this._rowCell = rowCell;
    return this;
  }

  public cornerCell(cornerCell: CellCallback) {
    this._cornerCell = cornerCell;
    return this;
  }

  public dataCell(dataCell: DataCellCallback) {
    this._dataCell = dataCell;
    return this;
  }

  public valueInCols(valueInCols: boolean) {
    this._valueInCols = valueInCols;
    return this;
  }

  public hideRowColFields(hideRowColFields: string[]) {
    this._hideRowColFields = hideRowColFields;
    return this;
  }

  public scrollReachNodeField(scrollReachNodeField: NodeField) {
    this._scrollReachNodeField = scrollReachNodeField;
    return this;
  }

  public registerDefaultInteractions(registerDefaultInteractions: boolean) {
    this._registerDefaultInteractions = registerDefaultInteractions;
    return this;
  }

  public keepOnlyNodesIds(keepOnlyNodesIds: KeepOnlyIds) {
    this._keepOnlyNodesIds = keepOnlyNodesIds;
    return this;
  }

  public hideNodesIds(hideNodesIds: string[]) {
    this._hideNodesIds = hideNodesIds;
    return this;
  }

  public showSeriesNumber(
    showSeriesNumber: boolean,
  ): SpreadSheetOptionsBuilder {
    this._showSeriesNumber = showSeriesNumber;
    return this;
  }

  public totals(totals: Totals): SpreadSheetOptionsBuilder {
    this._totals = totals;
    return this;
  }

  public style(style: Partial<Style>): SpreadSheetOptionsBuilder {
    this._style = style;
    return this;
  }

  public customHeaderCells(
    customHeaderCells: CustomHeaderCells,
  ): SpreadSheetOptionsBuilder {
    this._customHeaderCells = customHeaderCells;
    return this;
  }

  public spreadsheetType(spreadsheetType: boolean): SpreadSheetOptionsBuilder {
    this._spreadsheetType = spreadsheetType;
    return this;
  }

  public containsRowHeader(
    containsRowHeader: boolean,
  ): SpreadSheetOptionsBuilder {
    this._containsRowHeader = containsRowHeader;
    return this;
  }

  public pagination(pagination: Pagination): SpreadSheetOptionsBuilder {
    this._pagination = pagination;
    return this;
  }

  public linkFieldIds(linkFieldIds: string[]): SpreadSheetOptionsBuilder {
    this._linkFieldIds = linkFieldIds;
    return this;
  }

  public hierarchyType(
    hierarchyType: 'grid' | 'tree',
  ): SpreadSheetOptionsBuilder {
    this._hierarchyType = hierarchyType;
    return this;
  }

  public hierarchyCollapse(
    hierarchyCollapse: boolean,
  ): SpreadSheetOptionsBuilder {
    this._hierarchyCollapse = hierarchyCollapse;
    return this;
  }

  public height(height: number): SpreadSheetOptionsBuilder {
    this._height = height;
    return this;
  }

  public width(width: number): SpreadSheetOptionsBuilder {
    this._width = width;
    return this;
  }

  public conditions(conditions: Conditions): SpreadSheetOptionsBuilder {
    this._conditions = conditions;
    return this;
  }

  private _defaultStyle(): Style {
    return {
      treeRowsWidth: TREE_ROW_DEFAULT_WIDTH,
      collapsedRows: {},
      collapsedCols: {},
      cellCfg: {
        width: 96,
        height: 30,
        padding: 0,
      },
      rowCfg: {
        width: 96,
        widthByField: {},
      },
      colCfg: {
        height: 40,
        widthByFieldValue: {},
        heightByField: {},
        colWidthType: 'adaptive',
        totalSample: 10,
        detailSample: 30,
        showDerivedIcon: true,
        maxSampleIndex: 1,
      },
      plotSize: {
        width: 640,
        height: 480,
      },
      device: 'pc',
    } as Style;
  }
}
