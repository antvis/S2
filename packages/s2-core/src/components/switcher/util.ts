import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
