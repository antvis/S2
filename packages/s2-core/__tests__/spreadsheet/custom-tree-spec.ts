import { getContainer } from 'tests/util/helpers';
import { CustomTreePivotDataSet } from '../../src';
import type { HeaderCell } from '../../src/cell/header-cell';
import { customRowGridSimpleFields } from '../data/custom-grid-simple-fields';
import { customTreeNodes } from '../data/custom-tree-nodes';
import { CustomGridData } from '../data/data-custom-grid';
import {
  expectHighlightActiveNodes,
  getSelectedCount,
  getSelectedSum,
  getTestTooltipData,
} from '../util/interaction';
import { PivotSheet, SpreadSheet } from '@/sheet-type';
import type { S2DataConfig, S2Options } from '@/common/interface';

const s2Options: S2Options = {
  width: 600,
  height: 400,
  hierarchyType: 'tree',
  style: {
    rowCell: {
      width: 400,
    },
  },
};

describe('SpreadSheet Custom Tree Tests', () => {
  let s2: SpreadSheet;

  const getCornerCellLabels = () => {
    return s2.facet.getCornerCells().map((cell) => cell.getActualText());
  };

  const mapRowNodes = (spreadsheet: SpreadSheet) =>
    spreadsheet.facet.getRowLeafNodes().map((node) => {
      const iconName = (node.belongsCell as HeaderCell).getTreeIcon()?.config
        .name;

      return {
        id: node.id,
        value: node.value,
        isCollapsed: node.isCollapsed,
        iconName,
      };
    });

  const customRowDataCfg: S2DataConfig = {
    data: CustomGridData,
    meta: [
      {
        field: 'type',
        name: '类型',
      },
      {
        field: 'sub_type',
        name: '子类型',
      },
      {
        field: 'a-1',
        name: '层级1',
      },
      {
        field: 'a-2',
        name: '层级2',
      },
    ],
    fields: customRowGridSimpleFields,
  };

  beforeEach(async () => {
    s2 = new PivotSheet(getContainer(), customRowDataCfg, s2Options);
    await s2.render();
  });

  afterEach(() => {
    s2.destroy();
  });

  test('should disable valueInCols', () => {
    expect(s2.dataCfg.fields.valueInCols).toBeFalsy();
    expect(s2.dataSet.fields.valueInCols).toBeFalsy();
  });

  test('should use custom tree pivot dataSet', () => {
    expect(s2.dataSet).toBeInstanceOf(CustomTreePivotDataSet);
  });

  test('should get correctly dataset fields', () => {
    expect(s2.dataSet.fields).toMatchSnapshot();
  });

  test('should render custom layout row nodes', () => {
    const rowNodes = s2.facet.getRowNodes().map((node) => {
      return {
        value: node.value,
        width: node.width,
        height: node.height,
        description: node.extra?.['description'],
      };
    });

    expect(rowNodes).toMatchSnapshot();
  });

  test('should calc correctly row index of leaf nodes', () => {
    const rowLeafNodes = s2.facet.getRowLeafNodes().map((node) => {
      return {
        value: node.value,
        rowIndex: node.rowIndex,
      };
    });

    expect(rowLeafNodes).toMatchSnapshot();
  });

  test('should select custom row header cell', () => {
    // a-1
    const rowNode = s2.facet.getRowNodes()[0];

    // 选中 a-1
    s2.interaction.selectHeaderCell({
      cell: rowNode.belongsCell!,
    });

    // 选中单元格本身
    expect(s2.interaction.getActiveCells()).toHaveLength(1);
    // 高亮子节点
    expectHighlightActiveNodes(s2, ['root[&]a-1']);

    // 取消选中 a-1
    s2.interaction.selectHeaderCell({
      cell: rowNode.belongsCell!,
    });
    expect(s2.interaction.getActiveCells()).toBeEmpty();
  });

  test.each([
    { field: 'a-1', count: 2, sum: null },
    { field: 'a-1-1', count: 2, sum: null },
    { field: 'measure-1', count: 2, sum: 24 },
    { field: 'measure-2', count: 2, sum: 10 },
    { field: 'a-1-2', count: 2, sum: null },
    { field: 'a-1-2', count: 2, sum: null },
  ])(
    'should get selected cell summary infos for %o',
    ({ field, count, sum }) => {
      const rowNode = s2.facet
        .getRowNodes()
        .find((node) => node.field === field)!;

      // 选中
      s2.interaction.selectHeaderCell({
        cell: rowNode.belongsCell!,
      });

      const tooltipData = getTestTooltipData(s2, rowNode.belongsCell!);

      // 选中个数
      expect(getSelectedCount(tooltipData.summaries)).toEqual(count);
      // 汇总数据总和
      expect(getSelectedSum(tooltipData.summaries)).toEqual(sum);
    },
  );

  test('should render custom corner text by default title', async () => {
    s2.setDataCfg({
      meta: [],
    });

    await s2.render(true);

    const cornerCellLabels = getCornerCellLabels();

    expect(cornerCellLabels).toEqual([
      '自定义节点 a-1/自定义节点 a-2/数值',
      'type',
    ]);
  });

  test('should render custom corner text by meta formatter', async () => {
    s2.setDataCfg({
      meta: [
        {
          field: 'a-1',
          name: '文本1',
        },
        {
          field: 'a-2',
          name: '文本2',
        },
      ],
    });

    await s2.render(true);
    const cornerCellLabels = getCornerCellLabels();

    expect(cornerCellLabels).toEqual(['文本1/文本2/数值', 'type']);
  });

  test('should render custom corner text by cornerText options', async () => {
    s2.setOptions({
      cornerText: '测试',
    });

    await s2.render();

    const cornerCellLabels = getCornerCellLabels();

    expect(cornerCellLabels).toEqual(['测试', '类型']);
  });

  test('should render custom tree row node width', async () => {
    s2.setOptions({
      style: {
        rowCell: {
          width: 50,
        },
      },
    });

    await s2.render();

    const { rowNodes, rowsHierarchy } = s2.facet.getLayoutResult();

    expect(rowsHierarchy.width).toEqual(50);
    expect(rowNodes.every((node) => node.width === 50)).toBeTruthy();
  });

  test('should collapse node by collapsed', async () => {
    s2.setDataCfg({
      fields: {
        rows: customTreeNodes.map((node) => {
          return {
            ...node,
            collapsed: true,
          };
        }),
      },
    });
    await s2.render(true);

    expect(mapRowNodes(s2)).toMatchSnapshot();
  });

  test('should collapse node by collapsed fields id', async () => {
    s2.setOptions({
      style: {
        rowCell: {
          collapseFields: {
            'root[&]自定义节点 a-1': true,
            'root[&]自定义节点 a-2': false,
          },
        },
      },
    });
    await s2.render(true);

    expect(mapRowNodes(s2)).toMatchSnapshot();
  });

  test('should collapse node by collapsed field', async () => {
    s2.setOptions({
      style: {
        rowCell: {
          collapseFields: { 'a-1-1': true, 'a-1-2': false },
        },
      },
    });
    await s2.render(true);

    expect(mapRowNodes(s2)).toMatchSnapshot();
  });

  test('should collapse node by user collapseFields first', async () => {
    const collapsedField = 'custom-node-1';

    s2.setDataCfg({
      fields: {
        rows: customTreeNodes.map((node) => {
          return {
            ...node,
            collapsed: node.field !== collapsedField,
          };
        }),
      },
    });

    s2.setOptions({
      style: {
        rowCell: {
          collapseFields: {
            [collapsedField]: true,
          },
        },
      },
    });
    await s2.render(true);

    expect(mapRowNodes(s2)).toMatchSnapshot();
  });

  // https://github.com/antvis/S2/issues/2455
  test('should only collapse first node by node id', async () => {
    const collapsedField = 'custom-node-1';

    s2.setDataCfg({
      fields: {
        rows: customTreeNodes.map((node) => {
          return {
            ...node,
            // 让两个节点名一样
            title: '自定义节点A',
            collapsed: false,
          };
        }),
      },
    });

    s2.setOptions({
      style: {
        rowCell: {
          collapseFields: {
            [collapsedField]: true,
          },
        },
      },
    });
    await s2.render(true);

    expect(mapRowNodes(s2)).toMatchSnapshot();
  });
});
