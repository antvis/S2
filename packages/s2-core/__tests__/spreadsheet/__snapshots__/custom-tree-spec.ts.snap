// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpreadSheet Custom Tree Tests should calc correctly row index of leaf nodes 1`] = `
Array [
  Object {
    "rowIndex": 0,
    "value": "自定义节点 a-1",
  },
  Object {
    "rowIndex": 1,
    "value": "自定义节点 a-1-1",
  },
  Object {
    "rowIndex": 2,
    "value": "指标1",
  },
  Object {
    "rowIndex": 3,
    "value": "指标2",
  },
  Object {
    "rowIndex": 4,
    "value": "自定义节点 a-1-2",
  },
  Object {
    "rowIndex": 5,
    "value": "自定义节点 a-2",
  },
]
`;

exports[`SpreadSheet Custom Tree Tests should collapse node by collapsed 1`] = `
Array [
  Object {
    "iconName": "Plus",
    "id": "root[&]custom-node-1",
    "isCollapsed": true,
    "value": "自定义节点A",
  },
  Object {
    "iconName": "Plus",
    "id": "root[&]measure-e",
    "isCollapsed": true,
    "value": "指标E",
  },
]
`;

exports[`SpreadSheet Custom Tree Tests should collapse node by collapsed field 1`] = `
Array [
  Object {
    "iconName": "Minus",
    "id": "root[&]a-1",
    "isCollapsed": false,
    "value": "自定义节点 a-1",
  },
  Object {
    "iconName": "Plus",
    "id": "root[&]a-1[&]a-1-1",
    "isCollapsed": true,
    "value": "自定义节点 a-1-1",
  },
  Object {
    "iconName": undefined,
    "id": "root[&]a-1[&]a-1-2",
    "isCollapsed": false,
    "value": "自定义节点 a-1-2",
  },
  Object {
    "iconName": undefined,
    "id": "root[&]a-2",
    "isCollapsed": false,
    "value": "自定义节点 a-2",
  },
]
`;

exports[`SpreadSheet Custom Tree Tests should collapse node by collapsed fields id 1`] = `
Array [
  Object {
    "iconName": "Minus",
    "id": "root[&]a-1",
    "isCollapsed": false,
    "value": "自定义节点 a-1",
  },
  Object {
    "iconName": "Minus",
    "id": "root[&]a-1[&]a-1-1",
    "isCollapsed": false,
    "value": "自定义节点 a-1-1",
  },
  Object {
    "iconName": undefined,
    "id": "root[&]a-1[&]a-1-1[&]measure-1",
    "isCollapsed": false,
    "value": "指标1",
  },
  Object {
    "iconName": undefined,
    "id": "root[&]a-1[&]a-1-1[&]measure-2",
    "isCollapsed": false,
    "value": "指标2",
  },
  Object {
    "iconName": undefined,
    "id": "root[&]a-1[&]a-1-2",
    "isCollapsed": false,
    "value": "自定义节点 a-1-2",
  },
  Object {
    "iconName": undefined,
    "id": "root[&]a-2",
    "isCollapsed": false,
    "value": "自定义节点 a-2",
  },
]
`;

exports[`SpreadSheet Custom Tree Tests should collapse node by user collapseFields first 1`] = `
Array [
  Object {
    "iconName": "Plus",
    "id": "root[&]custom-node-1",
    "isCollapsed": true,
    "value": "自定义节点A",
  },
  Object {
    "iconName": "Plus",
    "id": "root[&]measure-e",
    "isCollapsed": true,
    "value": "指标E",
  },
]
`;

exports[`SpreadSheet Custom Tree Tests should only collapse first node by node id 1`] = `
Array [
  Object {
    "iconName": "Plus",
    "id": "root[&]custom-node-1",
    "isCollapsed": true,
    "value": "自定义节点A",
  },
  Object {
    "iconName": "Minus",
    "id": "root[&]measure-e",
    "isCollapsed": false,
    "value": "自定义节点A",
  },
  Object {
    "iconName": undefined,
    "id": "root[&]measure-e[&]custom-node-3",
    "isCollapsed": false,
    "value": "自定义节点C",
  },
  Object {
    "iconName": "Plus",
    "id": "root[&]measure-e[&]custom-node-4",
    "isCollapsed": true,
    "value": "自定义节点D",
  },
]
`;

exports[`SpreadSheet Custom Tree Tests should render custom layout row nodes 1`] = `
Array [
  Object {
    "description": "a-1 描述",
    "height": 30,
    "value": "自定义节点 a-1",
    "width": 400,
  },
  Object {
    "description": "a-1-1 描述",
    "height": 30,
    "value": "自定义节点 a-1-1",
    "width": 400,
  },
  Object {
    "description": "指标1描述",
    "height": 30,
    "value": "指标1",
    "width": 400,
  },
  Object {
    "description": "指标2描述",
    "height": 30,
    "value": "指标2",
    "width": 400,
  },
  Object {
    "description": "a-1-2 描述",
    "height": 30,
    "value": "自定义节点 a-1-2",
    "width": 400,
  },
  Object {
    "description": "a-2 描述",
    "height": 30,
    "value": "自定义节点 a-2",
    "width": 400,
  },
]
`;
